### ISSUE TITLE
ComboBox: factor out classes for subcomponents

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/sun#430]
ComboBox's list is currently just a Rectangle, and all of the responsibilities related to the list are spread out across ComboBox's constructor and methods.   Creating a `ListNode` inner class to handle list responsibilities would go a long ways towards cleaning up the PhET-iO (#405) and a11y (#314) instrumentation.

**Comment by pixelzoom** [comment]
@ariel-phet Permission to work on this?  It may be a multi-day project.

**Comment by ariel-phet** [comment]
@pixelzoom permission granted... ComboBox is used so much that having it in good shape seems very worthwhile for the project as whole.

**Comment by pixelzoom** [comment]
I think it might also be worthwhile to formalize "combo box list" by making it a class.  It's currently duck typed, and that makes it difficult to document.

**Comment by zepumph** [comment]
Could we make a new folder in sun for ComboBox related code? If there are 5 classes I would prefer that organization to having all of them in the same file. That may help with readability too. Though it would make a need for many require statements to be refactored.

**Comment by pixelzoom** [comment]
In light of #419, I'm rethinking the responsibilities and sharing of ComboBoxItemNode.   I think that what should be shared by the button and list is ComboBoxItem.node, and ComboBoxItemNode should become the thing that appears in the list, supports highlighting, etc.

**Comment by pixelzoom** [comment]
Over in https://github.com/phetsims/sun/issues/440, we decided that the combo box list class will be `ComboBoxListBox`, and in context:

```js
//ComboBoxListBox.js
ariaRole: 'listbox',

// ComboBox.js
this.listBox = new ComboBoxListBox( ..., { 
  tandem: tandem.createTandem( 'listBox' )
} );
```

**Comment by pixelzoom** [comment]
Closed accidentally, reopened.

**Comment by pixelzoom** [comment]
Summary of work done above:

* new classes `ComboBoxItem`, `ComboBoxListItemNode`, and `ComboBoxButton` 
* simplified the options API for `ComboBox`
* updated all clients to the new options API (not pixel-perfect, but close in all cases)

What I was _not_ able to do was factor out a "list box".   a11y made that pretty much impossible for me to accomplished, and I ultimately bailed.  That work is moved to https://github.com/phetsims/sun/issues/445.

@ariel-phet please assign someone to review the work that I've done here.

**Comment by pixelzoom** [comment]
I should also note that did the following tests of this work:

* Ran Beer's Law Lab in Studio. Comfirmed that all instrumented parts of ComboBox continue to work correctly.

* Ran Beer's Law Lab with `?a11y`.  Confirmed that keyboard navigation of ComboBox works.  I could navigate to the button (tab), pop up the listbox (spacebar), move between items in the list (arrow keys), and select an item in the list (spacebar).

**Comment by jessegreenberg** [comment]
@ariel-phet I could review this today, but I am also aware that @zepumph is doing some other combo box related work in #445 and #446, so review might be part of that work? But I am also happy to review, please reassign to me if you would like me to.

**Comment by ariel-phet** [comment]
@jessegreenberg actually I think this review should happen before that work. yes, please review today.

**Comment by pixelzoom** [comment]
Might be best to wait and do a comprehensive review (core implementation, a11y, phet-io) after #445 and #314 are done.

**Comment by pixelzoom** [comment]
Or as @ariel-phet said... :)    Do a review now, just to make sure nothing is seriously broken.

**Comment by jessegreenberg** [comment]
OK, starting with review!

I am seeing a few issues on CT that look like they could be related to recent ComboBox work:
<details> <summary>projectile-motion</summary>

```
projectile-motion : fuzz : require.js : load
Uncaught TypeError: this.projectileChoiceComboBox.hideListFromClick is not a function
TypeError: this.projectileChoiceComboBox.hideListFromClick is not a function
    at IntroProjectilePanel.hideComboBoxList (https://bayes.colorado.edu/continuous-testing/snapshot-1547660594488/projectile-motion/js/intro/view/IntroProjectilePanel.js?bust=1547664002190:202:37)
    at IntroScreenView.layout (https://bayes.colorado.edu/continuous-testing/snapshot-1547660594488/projectile-motion/js/intro/view/IntroScreenView.js?bust=1547664002190:67:33)
    at https://bayes.colorado.edu/continuous-testing/snapshot-1547660594488/joist/js/Sim.js?bust=1547664002190:124:18
    at u (https://bayes.colorado.edu/continuous-testing/snapshot-1547660594488/sherpa/lib/lodash-4.17.4.min.js?bust=1547664002190:5:432)
    at Function.ru (https://bayes.colorado.edu/continuous-testing/snapshot-1547660594488/sherpa/lib/lodash-4.17.4.min.js?bust=1547664002190:67:656)
    at Emitter.listener (https://bayes.colorado.edu/continuous-testing/snapshot-1547660594488/joist/js/Sim.js?bust=1547664002190:123:11)
    at Emitter.emit (https://bayes.colorado.edu/continuous-testing/snapshot-1547660594488/axon/js/Emitter.js?bust=1547664002190:218:55)
    at Sim.resize (https://bayes.colorado.edu/continuous-testing/snapshot-1547660594488/joist/js/Sim.js?bust=1547664002190:763:27)
    at Sim.resizeToWindow (https://bayes.colorado.edu/continuous-testing/snapshot-1547660594488/joist/js/Sim.js?bust=1547664002190:758:12)
    at Sim.finishInit (https://bayes.colorado.edu/continuous-testing/snapshot-1547660594488/joist/js/Sim.js?bust=1547664002190:706:12)
id: Bayes Chrome
Approximately 1/16/2019, 12:43:14 PM
```
</details>

<details> <summary>griddle, scenery-phet, wave-interference</summary>

```
griddle : fuzz : require.js-canvas : load
Uncaught Error: Assertion failed: Corner overlap on left edge
Error: Assertion failed: Corner overlap on left edge
    at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/snapshot-1547658484416/assert/js/assert.js:22:13)
    at Function.Shape.roundedRectangleWithRadii (https://bayes.colorado.edu/continuous-testing/snapshot-1547658484416/kite/js/Shape.js?bust=1547660391159:1952:15)
    at createButtonShape (https://bayes.colorado.edu/continuous-testing/snapshot-1547658484416/sun/js/buttons/RectangularButtonView.js?bust=1547660391159:38:18)
    at ComboBoxButton.RectangularButtonView [as constructor] (https://bayes.colorado.edu/continuous-testing/snapshot-1547658484416/sun/js/buttons/RectangularButtonView.js?bust=1547660391159:115:28)
    at new RectangularPushButton (https://bayes.colorado.edu/continuous-testing/snapshot-1547658484416/sun/js/buttons/RectangularPushButton.js?bust=1547660391159:47:27)
    at new ComboBoxButton (https://bayes.colorado.edu/continuous-testing/snapshot-1547658484416/sun/js/ComboBoxButton.js?bust=1547660391159:111:7)
    at new ComboBox (https://bayes.colorado.edu/continuous-testing/snapshot-1547658484416/sun/js/ComboBox.js?bust=1547660391159:262:19)
    at new DemosScreenView (https://bayes.colorado.edu/continuous-testing/snapshot-1547658484416/sun/js/demo/DemosScreenView.js?bust=1547660391159:88:20)
    at new GriddleDemoScreenView (https://bayes.colorado.edu/continuous-testing/snapshot-1547658484416/griddle/js/demo/GriddleDemoScreenView.js?bust=1547660391159:43:7)
    at Screen.name [as createView] (https://bayes.colorado.edu/continuous-testing/snapshot-1547658484416/griddle/js/griddle-main.js?bust=1547660391159:27:63)
id: Bayes Chrome
Approximately 1/16/2019, 12:08:04 PM
```
</details>

**Comment by pixelzoom** [comment]
Those issues have been addressed in https://github.com/phetsims/projectile-motion/issues/170 and https://github.com/phetsims/projectile-motion/issues/171.

**Comment by jessegreenberg** [comment]
I reviewed the changes. The new classes `ComboBoxButton`, `ComboBoxListItemNode` and `ComboBoxItem` all look very nice and make sense to me.  I tested ComboBox options in the sun demo and all worked as I expected. In reviewing I only noticed two things:

1) The look in sims slightly changed because the option defaults and behavior for margins changed and the default for `cornerRadius` was reduced by half. I reviewed bending-light and molecule-shapes and the differences seem OK to me but want to make sure this is acceptable.
2) Clicking the ComboBoxButton when the list is shown immediately opens the box again which didn't happen before the changes. This is probably caused by the removal of `enableClickToDismissListener` (https://github.com/phetsims/scenery/issues/58#issuecomment-454503450) and may be more directly related to work in https://github.com/phetsims/sun/issues/446.

This is a big improvement to ComboBox @pixelzoom, thanks! Back to you.

**Comment by pixelzoom** [comment]
> 1. The look in sims slightly changed ... want to make sure this is acceptable.

This is a question for @ariel-phet. 

FYI, trying to make things look exact is a rathole that I did go down.  But there were so many things that made that difficult to accomplish.  There were something about layout that were just wrong to begin with inside ComboBox and its components.  The options for margins were overly complicated and were changed dramatically.  And some of the defaults didn't really make senses as defaults; they were probably the values used by the sim that birthed ComboBox (BLL?)   Anyway... I examined every ComboBox in every sim (which took forever) and verified that they all look reasonable.  (Some btw did not look great before the changes.)  If we want everything to look exactly the same as it did before the changes, that's going to take an undeterminable amount of time.

> 2. Clicking the ComboBoxButton when the list is shown immediately opens the box again which didn't happen before the changes. This is probably caused by the removal of `enableClickToDismissListener` (phetsims/scenery#58 (comment)) and may be more directly related to work in #446.

This can be fixed by making the button listener toggle the visibility of the list.  I'll handle that now.

**Comment by pixelzoom** [comment]
Hmmm... I thought that addressing (2) should have been as simple as:

```js
    this.button.addListener( () => {
      console.log( 'this.listNode.visible=' + this.listNode.visible );//XXX
      if ( !this.listNode.visible ) {
        this.showList();
      }
      else {
        this.hideList();
      }
    } );
```

... but while the list is visible, `clickToDismissListener` is getting the event before the button, and it does this:

```js
    this.clickToDismissListener = {
      down: () => { this.hideList(); }
    };
```

So `clickToDismissListener` hides the list, and the button's listener then shows it.  Should `clickToDismissListener` be doing something to ensure that the event is not propagated to the button?

**Comment by jessegreenberg** [comment]
I tried `event.abort()` in `clickToDismissListener` but it didn't work because apparently the display listeners are called *after* the event is dispatched to nodes in the trail. So the `down` listener of the ComboBoxButton is called before `clickToDismissListener`.

**Comment by pixelzoom** [comment]
Yes @jessegreenberg, sounds like we tried the same things with `event.abort`.  I have some other ideas, will investigate more tomorrow.

**Comment by pixelzoom** [comment]
@jessegreenberg Actually, I see `clickToDismissListener`'s `down` function being called _before_ the `this.button`'s listener.

**Comment by pixelzoom** [comment]
In the previous commit, clicking on the button toggles the list's visibility:

```js
    // Clicking on the button toggles visibility of the list
    this.button.addListener( () => {
      if ( !this.listBox.visible ) {
        this.showList();
      }
      else {
        this.hideList();
      }
    } );
```

And `clickToDismissListener` is short-circuited when the click is over the button:

```js
    this.clickToDismissListener = {
      down: event => {

        //TODO sun#430 is Trail.nodes public? Is there a better way to determine if we clicked over this.button?
        // Ignore if we click over the button, since the button will handle hiding the list.
        if ( event.trail.nodes.indexOf( this.button ) === -1 ) {
          this.hideList();
        }
      }
    };
```

This seems to do the job.  But see the TODO above.`Trail` does not have visibility annotations for its fields, so I don't know if `event.trail.nodes` is `@public`. (I've asked @jonathanolson on Slack.) And I'm surprised that Trail doesn't have a method that determines whether a specific Node is part of the  Trail.

**Comment by jessegreenberg** [comment]
> Actually, I see clickToDismissListener's down function being called before the this.button's listener.

That is right, because the button's listener is called on `up`. I was trying to prevent the button from being pressed down in the first place.

**Comment by pixelzoom** [comment]
Blocked by https://github.com/phetsims/scenery/issues/927.

**Comment by pixelzoom** [comment]
I see that other sims are accessing `trail.nodes`, doing the same sort of test that I'm doing here in ComboBox.  So let's proceed with this solution.   If @jonathanolson decides in https://github.com/phetsims/scenery/issues/927 that `trail.nodes` should be private, then a method to perform the test will be added to Trail.  So...

@jessegreenberg regarding your feedback in https://github.com/phetsims/sun/issues/430#issuecomment-454984350... 

(1) is being addressed by https://github.com/phetsims/sun/issues/452.

(2) has been addressed as described in https://github.com/phetsims/sun/issues/430#issuecomment-455037261.

Anything else to do here? If not, feel free to close this issue.

**Comment by jessegreenberg** [comment]
OK, thanks  for going through https://github.com/phetsims/sun/issues/452. I tested https://github.com/phetsims/sun/issues/430#issuecomment-455037261, it looks good and works good. I am going to remove this TODO now:
> //TODO sun#430 is Trail.nodes public? Is there a better way to determine if we clicked over this.button?

Closing.

**Comment by pixelzoom** [comment]
Thanks for the review @jessegreenberg.

FYI, I added this TODO where you removed the sun#430 TODO:

`//TODO scenery#927 is trail.nodes public? should we add Trail.containsNode?`
